lcd.lua

```diff
@@ -30,31 +30,88 @@ local NUMBER_OF_LINES = 5
 local LINE_HEIGHT = 14
 local CHAR_WIDTH = 5
 
+local split = function(s, pat)
+	-- adapted from https://stackoverflow.com/a/1647577/4067384
+	pat = pat or '%s+'
+	local st, g = 1, string.gmatch(s, "()("..pat..")")
+	local function getter(segs, seps, sep, cap1, ...)
+		st = sep and seps + #sep
+		return string.sub(s, segs, (seps or 0) - 1), cap1 or sep, ...
+	end
+	return function() if st then return getter(st, g()) end end
+end
```

Comment:
This function, as far as I can tell, is correct, but seems overly complicated. It uses variadic parameters even though there will only ever be two pattern captures (and even `cap1` will always be `nil` and is also never used for anything), it returns multiple values even though only the first return value is ever used, it (IMO confusingly) passes `st` as a parameter named `segs` but then also refers to `st` inside the body (why not just not bother with the parameter and use `st` via the closure in both cases?). It also provides a default pattern even though both calls to it provide explicit patterns, and neither is even the same as the default.

What’s wrong with this? AFAICT it’s simpler and also potentially faster because it eliminates the extra level of function call (the anonymous lambda calling `getter` no longer exists, instead it just returns `getter` directly).
```lua
local split = function(s, pat)
	-- adapted from https://stackoverflow.com/a/1647577/4067384
	local st, g = 1, s:gmatch("()("..pat..")")
	local function getter()
		if st then
			local segs, seps, sep = st, g()
			st = sep and seps + #sep
			return s:sub(segs, (seps or 0) - 1)
		end
	end
	return getter
end
```

Reply:
I just took the stackoverflow solution without diving into the how, it worked, and I left it at that. Feel free to change.